keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,960 @@
+4. Non-goal: Create a new empty group and in the same time create new PVCs and add to the new group.
+
+Modify an existing VolumeGroup:
+Add new volume or remove existing volume from an existing VolumeGroup. Option 2 for create VolumeGroup above falls into this case.
+
+Two types of VolumeGroups can be supported:
+
+Type 1: Mutable VolumeGroup - Create new PVC and add to the group.
+* Create a new empty VolumeGroup.
+* Create a new PVC with existing VolumeGroup name. As a result, new PVC is created and added to VolumeGroup. VolumeGroup is modified so Status has this new PVC in PVCList.
+* External-provisioner will be modified so that VolumeGroupName will be passed to the CSI driver when creating a volume.
+
+Only CSI drivers supporting VOLUMEGROUP capability can support this type of mutable volume group.
+When a new PVC is created with the existing VolumeGroup name, the VolumeGroup will be modified and the PVC will be added to PVCList in the Status.
+
+Type 2: Mutable VolumeGroup - Add existing PVC to the group or remove PVC from the group without deleting it. A VOLUMEGROUP_ADD_REMOVE_EXISTING_VOLUME capability will be added to CSI Spec. Only CSI drivers supporting both VOLUMEGROUP and VOLUMEGROUP_ADD_REMOVE_EXISTING_VOLUME capabilities can support this type of mutable volume group.
+* Type 2 is a subset of Type 1 VolumeGroup.
+* Create a new empty VolumeGroup. VolumeGroup Status has a boolean AddRemoveExistingPVC set to true.
+* Add an existing PVC to an existing VolumeGroup (VolumeGroup can be empty to start with or it can have other PVCs already) by adding VolumeGroup name to the PVC Spec.
+* VolumeGroup is modified so the existing PVC is added to the PVCList in the Status.
+  * Note: The VolumeGroup controller will be implemented to have a desired state
+    of the world and an actual state of the world. The desired state of the world
+    contains VolumeGroups with the desired PVCList while the actual state of the
+    world contains VolumeGroups with the actual PVCList. The controller will try
+    to reconcile the two by handling adding and removing multiple PVCs through a
+    single CSI RPC call each time.
+* External-provisioner will be modified so that modifying PVC by adding VolumeGroupName will trigger a ModifyVolume call (a new CSI controller RPC) to CSI driver.
+* VolumeGroup controller will be triggered to update the VolumeGroup Status.
+* Deleting a PVC from a VolumeGroup will trigger external-provisioner and the VolumeGroup controller as well.
+
+CSI drivers that do not have a volume_group_id concept can use the VolumeGroup name stored in Kubernetes API server as the volume_group_id.
+
+Create Snapshot of a VolumeGroup:
```

Comment:
Is it allowed to create volume snapshots (using CreateSnapshot CSI gRPC) for volumes that belong to a VolumeGroup? Or, is it that once a volume belongs to a volume group only group snapshots are allowed?

I ask to understand how to overcome specific storage provider implementations where child snapshots are disallowed if parents snapshots exist (where a child would be a volume and the parent would be the volume group), or if such a restrictions by the storage provider is not supported by this specification.

Reply:
I think you should be able to create volume snapshot for volumes belonging to a VolumeGroup.  Are you aware of any storage provider that cannot support this?

Reply:
For CephFS we recently made some changes to support snapshot scalability, that limits the possibility of nested snapshots. Subvolume groups construct already exists in CephFS, and in reality the group is a parent subdirectory of the volumes in the volume group. Hence a snapshot at the group level would prevent subvolume/subdirectory level snapshots as nested snapshots are prevented.

It does seem correct from a specification perspective that independent snapshot lifecycle for volumes should not be limited just because the volumes belong to a group. So, attempting to understand/ensure that this is the case, and hence derive what work maybe needed with CephFS to support this in the future.

cc @batrick

Reply:
Maybe we need to add a capability for supporting taking a snapshot of a volume already in a group.

Reply:
> For CephFS we recently made some changes to support snapshot scalability, that limits the possibility of nested snapshots. Subvolume groups construct already exists in CephFS, and in reality the group is a parent subdirectory of the volumes in the volume group. Hence a snapshot at the group level would prevent subvolume/subdirectory level snapshots as nested snapshots are prevented.
> 
> It does seem correct from a specification perspective that independent snapshot lifecycle for volumes should not be limited just because the volumes belong to a group. So, attempting to understand/ensure that this is the case, and hence derive what work maybe needed with CephFS to support this in the future.
> 
> cc @batrick

I actually don't think there is a problem with snapshotting a subvolume group (in my limited testing) while allowing subvolumes to have independent and disjoint snapshots. I need to check the recent changes Zheng made to verify. So, this shouldn't be a problem for us actually.

Reply:
As discussed in last review meeting, I added a CSI capability INDIVIDUAL_SNAPSHOT_RESTORE to indicate whether a CSI driver can support creating a volume from an individual volume snapshot if the volume snapshot is part of a VolumeGroupSnapshot.